felix/nftables/table.go

```diff
@@ -0,0 +1,1173 @@
+	// rules with unknown hashes.
+	chainToInsertedRules map[string][]generictables.Rule
+
+	// chainToAppendRules maps from chain name to a list of rules to be appended at the end
+	// of that chain.
+	chainToAppendedRules map[string][]generictables.Rule
+
+	dirtyBaseChains set.Set[string]
+
+	// chainNameToChain contains the desired state of our chains, indexed by
+	// chain name.
+	chainNameToChain map[string]*generictables.Chain
+
+	// chainRefCounts counts the number of chains that refer to a given chain.  Transitive
+	// reachability isn't tracked but testing whether a chain is referenced does allow us to
+	// avoid programming unreferenced leaf chains (for example, policies that aren't used in
+	// this table).
+	chainRefCounts map[string]int
+	dirtyChains    set.Set[string]
+
+	inSyncWithDataPlane bool
+
+	// chainToDataplaneHashes contains the rule hashes that we think are in the dataplane.
+	// it is updated when we write to the dataplane but it can also be read back and compared
+	// to what we calculate from chainToContents.
+	chainToDataplaneHashes map[string][]string
+
+	// chainToFullRules contains the full rules for any chains that we may be hooking into, mapped from chain name
+	// to slices of rules in that chain.
+	chainToFullRules map[string][]*knftables.Rule
+
+	// hashCommentPrefix holds the prefix that we prepend to our rule-tracking hashes.
+	hashCommentPrefix string
+
+	// ourChainsRegexp matches the names of chains that belong to this specicific table.
+	ourChainsRegexp *regexp.Regexp
+
+	// Record when we did our most recent reads and writes of the table.  We use these to
+	// calculate the next time we should force a refresh.
+	lastReadTime             time.Time
+	lastWriteTime            time.Time
+	initialPostWriteInterval time.Duration
+	postWriteInterval        time.Duration
+	refreshInterval          time.Duration
+
+	// Estimates for the time taken to do an nftables read / write.
+	// When an nft command exceeds the one of these we update them immediately.
+	// When a  nft command takes less time we decay them exponentially.
+	peakNftablesReadTime  time.Duration
+	peakNftablesWriteTime time.Duration
+
+	logCxt               *log.Entry
+	updateRateLimitedLog *logutilslc.RateLimitedLogger
+
+	gaugeNumChains        prometheus.Gauge
+	gaugeNumRules         prometheus.Gauge
+	countNumLinesExecuted prometheus.Counter
+
+	// Factory for making commands, used by UTs to shim exec.Command().
+	newCmd cmdshim.CmdFactory
+
+	// Shims for time.XXX functions:
+	timeSleep func(d time.Duration)
+	timeNow   func() time.Time
+
+	onStillAlive func()
+	opReporter   logutils.OpRecorder
+	reason       string
+
+	contextTimeout time.Duration
+}
+
+type TableOptions struct {
+	// NewDataplane is an optional function to override the creation of the knftables client,
+	// used for testing.
+	NewDataplane func(knftables.Family, string) (knftables.Interface, error)
+
+	RefreshInterval   time.Duration
+	PostWriteInterval time.Duration
```

Comment:
I ended up ripping out the postWrite logic from nftables since it seems pretty unnecessary here. Can always revisit adding it via a common implementation if we need it!